static, monitoring: add documents for Longhorn monitoring system (#195)
longhorn/website: content/docs/1.1.0/monitoring/prometheus_and_grafana_setup.md

* Prometheus server which scrapes and stores time series data from Longhorn metrics endpoints. The Prometheus is also responsible for generating alerts base on configured rules and collected data. Prometheus servers then send alerts to an Alertmanager.
* AlertManager then manages those alerts, including silencing, inhibition, aggregation, and sending out notifications via methods such as email, on-call notification systems, and chat platforms.
* Grafana which queries Prometheus server for data and draws a dashboard for visualization.

The below picture describes the detailed architecture of the monitoring system.

![images](/img/screenshots/monitoring/longhorn-monitoring-system.png)

There are 2 unmentioned components in the above picture:

* Longhorn Backend service is a service pointing to the set of Longhorn manager pods. Longhorn's metrics are exposed in Longhorn manager pods at the endpoint `http://LONGHORN_MANAGER_IP:PORT/metrics`.

> * [Prometheus operator](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/user-guides/getting-started.md) make running Prometheus on top of Kubernetes become very easy. It watches 3 custom resource definitions: ServiceMonitor CRD, Prometheus CRD, AlertManager CRD. When users create those CRDs, Prometheus Operator deploys and manages the Prometheus server, AlerManager with the user-specified configurations.

Reviewer: ```suggestion
* [Prometheus operator](https://github.com/prometheus-operator/prometheus-operator/blob/master/Documentation/user-guides/getting-started.md) makes running Prometheus on top of Kubernetes very easy. The operator watches 3 custom resources: ServiceMonitor, Prometheus and AlertManager.
When users create those custom resources, Prometheus Operator deploys and manages the Prometheus server, AlerManager with the user-specified configurations.
```

Author: Thank you. Fixed

## Installation

> Following this instruction will install all components into the `monitoring` namespace. To install them to a different namespace, change the field `namespace: OTHER_NAMESPACE`

Reviewer: ```suggestion
Following this instruction will install all components into the `monitoring` namespace. To install them into a different namespace, change the field `namespace: OTHER_NAMESPACE`
```

Author: Thank you. Fixed

        ports:
        - containerPort: 8080
          name: http
        resources:
          limits:
            cpu: 200m
            memory: 200Mi
          requests:
            cpu: 100m
            memory: 100Mi
        securityContext:
          allowPrivilegeEscalation: false
      nodeSelector:
        beta.kubernetes.io/os: linux
      securityContext:
        runAsNonRoot: true
        runAsUser: 65534
      serviceAccountName: prometheus-operator
---
apiVersion: v1
kind: ServiceAccount
metadata:
  labels:
    app.kubernetes.io/component: controller
    app.kubernetes.io/name: prometheus-operator
    app.kubernetes.io/version: v0.38.3
  name: prometheus-operator
  namespace: monitoring
---
apiVersion: v1
kind: Service
metadata:
  labels:
    app.kubernetes.io/component: controller
    app.kubernetes.io/name: prometheus-operator
    app.kubernetes.io/version: v0.38.3
  name: prometheus-operator
  namespace: monitoring
spec:
  clusterIP: None
  ports:
  - name: http
    port: 8080
    targetPort: http
  selector:
    app.kubernetes.io/component: controller
    app.kubernetes.io/name: prometheus-operator
```

### Install Longhorn ServiceMonitor

> Longhorn ServiceMonitor has a label selector `app: longhorn-manager` to select Longhorn backend service. Later on, the Prometheus CRD can include Longhorn ServiceMonitor so that the Prometheus server can discover all Longhorn manager pods and their endpoints.

Reviewer: ```suggestion
Longhorn ServiceMonitor has a label selector `app: longhorn-manager` to select Longhorn backend service. 
Later on, the Prometheus custom resource can include Longhorn ServiceMonitor so that the Prometheus server can discover all Longhorn manager pods and their endpoints.
```
Don't know what this sentce means?

Author: Sorry for the confusion. I mean Longhorn ServiceMonitor has a label selector `app: longhorn-manager`  to select Longhorn manager backend service. Later on, Prometheus CRD selects  Longhorn ServiceMonitor. So the chain looks like this:
```
Longhorn manager pods <- Longhorn backend service <- Longhorn servicemonitor <- Prometheus
```

```yaml
apiVersion: monitoring.coreos.com/v1
kind: ServiceMonitor
metadata:
  name: longhorn-prometheus-servicemonitor
  namespace: monitoring
  labels:
    name: longhorn-prometheus-servicemonitor
spec:
  selector:
    matchLabels:
      app: longhorn-manager
  namespaceSelector:
    matchNames:
    - longhorn-system
  endpoints:
  - port: manager
```

### Install and configure Prometheus AlertManager

1. Create a highly available Alertmanager deployment with 3 instances:

    ```yaml
    apiVersion: monitoring.coreos.com/v1
    kind: Alertmanager
    metadata:
      name: longhorn
      namespace: monitoring
    spec:
      replicas: 3
    ```
1. The Alertmanager instances will not be able to start up unless a valid configuration is given.

> See more explanation about Alertmanager configuration at [here](https://prometheus.io/docs/alerting/latest/configuration/).

Reviewer: ```suggestion
See [here](https://prometheus.io/docs/alerting/latest/configuration/) for more information about Alertmanager configuration.
```

Author: Thank you. Fixed

### Install and configure Prometheus server

1. Create PrometheusRule custom resource which defines alert conditions. See more examples about Longhorn alert rules at [Longhorn Alert Rule Examples](../alert-rules-example).

    ```yaml
    apiVersion: monitoring.coreos.com/v1
    kind: PrometheusRule
    metadata:
      labels:
        prometheus: longhorn
        role: alert-rules
      name: prometheus-longhorn-rules
      namespace: monitoring
    spec:
      groups:
      - name: longhorn.rules
        rules:
        - alert: LonghornVolumeUsageCritical
          annotations:
            description: Longhorn volume {{$labels.volume}} on {{$labels.node}} is at {{$value}}% used for
              more than 5 minutes.
            summary: Longhorn volume capacity is over 90% used.
          expr: 100 * (longhorn_volume_usage_bytes / longhorn_volume_capacity_bytes) > 90
          for: 5m
          labels:
            issue: Longhorn volume {{$labels.volume}} usage on {{$labels.node}} is critical.
            severity: critical
    ```

>    See more about how to define alert rules at [here](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/#alerting-rules).

Reviewer: ```suggestion
   For more information on how to define alert rules see [here](https://prometheus.io/docs/prometheus/latest/configuration/alerting_rules/#alerting-rules).
```

Author: Thank you. Fixed

1. If [RBAC](https://kubernetes.io/docs/reference/access-authn-authz/authorization/) authorization is activated, Create a ClusterRole and ClusterRoleBinding for the Prometheus Pods:

    ```yaml
    apiVersion: v1
    kind: ServiceAccount
    metadata:
      name: prometheus
      namespace: monitoring
    ```

    ```yaml
    apiVersion: rbac.authorization.k8s.io/v1beta1
    kind: ClusterRole
    metadata:
      name: prometheus
      namespace: monitoring
    rules:
    - apiGroups: [""]
      resources:
      - nodes
      - services
      - endpoints
      - pods
      verbs: ["get", "list", "watch"]
    - apiGroups: [""]
      resources:
      - configmaps
      verbs: ["get"]
    - nonResourceURLs: ["/metrics"]
      verbs: ["get"]
    ```

    ```yaml
    apiVersion: rbac.authorization.k8s.io/v1beta1
    kind: ClusterRoleBinding
    metadata:
      name: prometheus
    roleRef:
      apiGroup: rbac.authorization.k8s.io
      kind: ClusterRole
      name: prometheus
    subjects:
    - kind: ServiceAccount
      name: prometheus
      namespace: monitoring
    ```

> 1. Create Prometheus CRD. Notice that we select the Longhorn service monitor and Longhorn rules in Prometheus CRD's spec.

Reviewer: ```suggestion
1. Create a Prometheus custom resource. Notice that we select the Longhorn service monitor and Longhorn rules in the spec.
```

Author: Thank you. Fixed
